File: Design_Patterns/Observer_Design_pattern/java/Observable/WeatherStation.java
package Observable;

import Observer.Observer;
import java.util.ArrayList;
import java.util.List;

/**
 * WeatherStation - Concrete Observable Implementation
 * 
 * This class represents a weather station that monitors temperature.
 * When the temperature changes, it notifies all registered observers
 * (like Mobile and TV displays).
 * 
 * This demonstrates the Observer Design Pattern where:
 * - WeatherStation is the Subject/Observable
 * - It maintains a list of observers
 * - When temperature changes, all observers are automatically notified
 */
public class WeatherStation implements Observable {
    private List<Observer> observers;
    private int temperature;

    public WeatherStation() {
        this.observers = new ArrayList<>();
        this.temperature = 0;
    }

    @Override
    public void add(Observer observer) {
        observers.add(observer);
        System.out.println("Observer added. Total observers: " + observers.size());
    }

    @Override
    public void remove(Observer observer) {
        observers.remove(observer);
        System.out.println("Observer removed. Total observers: " + observers.size());
    }

    @Override
    public void notifyObservers() {
        System.out.println("\n--- Notifying all observers ---");
        for (Observer observer : observers) {
            observer.update();
        }
        System.out.println("--- Notification complete ---\n");
    }

    /**
     * Set new temperature and notify all observers
     */
    public void setTemperature(int newTemperature) {
        System.out.println("\nWeatherStation: Temperature changed from " +
                temperature + "°C to " + newTemperature + "°C");
        this.temperature = newTemperature;
        notifyObservers();
    }

    /**
     * Get current temperature
     */
    public int getTemperature() {
        return temperature;
    }
}

File: Design_Patterns/BehaviouralDesignPattern/Observer_Design_pattern/cpp/WeatherStation.cpp
/**
 * WeatherStation.cpp
 * 
 * Implementation of WeatherStation class
 */

#include "WeatherStation.h"
#include <iostream>
#include <algorithm>

WeatherStation::WeatherStation() : temperature(0) {}

void WeatherStation::add(Observer* observer) {
    observers.push_back(observer);
    std::cout << "Observer added. Total observers: " << observers.size() << std::endl;
}

void WeatherStation::remove(Observer* observer) {
    auto it = std::find(observers.begin(), observers.end(), observer);
    if (it != observers.end()) {
        observers.erase(it);
    }
    std::cout << "Observer removed. Total observers: " << observers.size() << std::endl;
}

void WeatherStation::notifyObservers() {
    std::cout << "\n--- Notifying all observers ---" << std::endl;
    for (Observer* observer : observers) {
        observer->update();
    }
    std::cout << "--- Notification complete ---\n" << std::endl;
}

void WeatherStation::setTemperature(int newTemperature) {
    std::cout << "\nWeatherStation: Temperature changed from " 
              << temperature << "°C to " << newTemperature << "°C" << std::endl;
    temperature = newTemperature;
    notifyObservers();
}

int WeatherStation::getTemperature() const {
    return temperature;
}
